batched/dense/unit_test/Test_Batched_SerialPttrf.hpp

```diff
@@ -0,0 +1,384 @@
+                   BlkSize),  // Diagonal components
+      ones(Kokkos::view_alloc("ones", Kokkos::WithoutInitializing), N, BlkSize);
+  View2DType e(Kokkos::view_alloc("e", Kokkos::WithoutInitializing), N,
+               BlkSize - 1);  // Upper and lower diagonal components (identical)
+
+  Kokkos::deep_copy(d, RealType(4.0));
+  Kokkos::deep_copy(e, ScalarType(1.0));
+  Kokkos::deep_copy(ones, RealType(1.0));
+
+  // Reconstruct Tridiaonal matrix A
+  // A = D + EL + EU
+  create_diagonal_matrix(e, EL, -1);
+  create_diagonal_matrix(e, EU, 1);
+  create_diagonal_matrix(d, D);
+  create_diagonal_matrix(ones, I);
+
+  // Matrix matrix addition by Gemm
+  // D + EU by D * I + EU (result stored in EU)
+  Functor_BatchedSerialGemm<DeviceType, ScalarType, View3DType, View3DType,
+                            View3DType, Trans::NoTranspose>(1.0, D, I, 1.0, EU)
+      .run();
+
+  // Copy EL to A
+  Kokkos::deep_copy(A, EL);
+
+  // EU + EL by EU * I + A (result stored in A)
+  Functor_BatchedSerialGemm<DeviceType, ScalarType, View3DType, View3DType,
+                            View3DType, Trans::NoTranspose>(1.0, EU, I, 1.0, A)
+      .run();
+
+  // Factorize matrix A -> L * D * L**T
+  // d and e are updated by pttrf
+  Functor_BatchedSerialPttrf<DeviceType, RealView2DType, View2DType,
+                             AlgoTagType>(d, e)
+      .run();
+
+  Kokkos::fence();
+
+  // Reconstruct L and D from factorized matrix
+  create_diagonal_matrix(e, EL, -1);
+  create_diagonal_matrix(d, D);
+
+  // Copy I to L
+  Kokkos::deep_copy(L, I);
+
+  // EL + I by EL * I + L (result stored in L)
+  Functor_BatchedSerialGemm<DeviceType, ScalarType, View3DType, View3DType,
+                            View3DType, Trans::NoTranspose>(1.0, EL, I, 1.0, L)
+      .run();
+
+  // Reconstruct A by L*D*L**T
+  // Gemm to compute L*D -> LD
+  Functor_BatchedSerialGemm<DeviceType, ScalarType, View3DType, View3DType,
+                            View3DType, Trans::NoTranspose>(1.0, L, D, 0.0, LD)
+      .run();
+
+  // Gemm to compute (L*D)*L**T -> A_reconst
+  Functor_BatchedSerialGemm<DeviceType, ScalarType, View3DType, View3DType,
+                            View3DType, Trans::Transpose>(1.0, LD, L, 0.0,
+                                                          A_reconst)
+      .run();
+
+  Kokkos::fence();
+
+  // this eps is about 10^-14
+  RealType eps = 1.0e3 * ats::epsilon();
+
+  auto h_A = Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), A);
+  auto h_A_reconst =
+      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), A_reconst);
+
+  // Check A = L*D*L.T
+  for (int ib = 0; ib < N; ib++) {
+    for (int i = 0; i < BlkSize; i++) {
+      for (int j = 0; j < BlkSize; j++) {
+        EXPECT_NEAR_KK(h_A_reconst(ib, i, j), h_A(ib, i, j), eps);
+      }
+    }
+  }
+}
+
+}  // namespace Pttrf
+}  // namespace Test
+
+template <typename DeviceType, typename ScalarType, typename AlgoTagType>
+int test_batched_pttrf() {
```

Comment:
I would recommend testing special cases like matrix with zero rows or matrix with only diagonal entries, etc...

Reply:
Do you recommend to add tests for (1x1) and (0x0) matrices?
I am not quite sure what should I do for these cases.
I can modify a function to disallow these and add a test to confirm that errors are raised in debug mode

Reply:
I have tested this function with lapack. 
I found that (1x1) and (0x0) are valid, but they just return the original value.
I have updated the code to follow this behavior. 

```C++
#include <iostream>
#include <vector>

extern "C" int dpttrf_(int const* n, double* d, double* e, int* info);

int main(int argc, char** argv){
    int info;

    std::vector<double> d, e;
    d.push_back(4.0);
    int N = d.size();

    dpttrf_(&N, d.data(), e.data(), &info);
    std::cout << "d (0) " << d.at(0) << std::endl; 
    std::cout << "Info = " << info << std::endl;

    std::vector<double> d0, e0;
    int N0 = 0;

    dpttrf_(&N0, d0.data(), e0.data(), &info);
    std::cout << "Info = " << info << std::endl; 

    return(0);
}
```

This returns.

```bash
d (0) 4
Info = 0
Info = 0
```